### PR TITLE
Add SSD1305 support to SSD1306 integration along with few new options

### PR DESCRIPTION
**Pull request by zhangjingye03**
# What does this implement/fix? 

* Added SSD1305 support to SSD1306 integration.  
    SSD1305 is similar to SSD1306, it seems SSD1305 has brightness and color register but does not have charge pump. The SSD1306 integration somehow uses `brightness` variable in yaml configuration but writes its value to `contrast` register `0x81` because there is no `brightness` register `0x82` in SSD1306. This PR fixes the wrong variable name, exposes both `brightness` and `contrast` to SSD1305.
    However, users using `brightness` option need to change it to `contrast` to get it work on SSD1306/SH1106. A warning would be raised if `brightness  != 1.0` is detected.
* Added some description when manipulating registers.
* Added `flip_x`, `flip_y`, `offset_x`, `offset_y` and `invert` option.
    My SSD1305 module got from AliExpress needs to shift X column by 4 pixels to get all things worked correctly. These options would show more compatibility to various display modules.
* Fixed typo `setup_ssd1036` -> `setup_ssd1306`.
* Tested on ESP8266 with SSD1305 connected via I2C.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1237

## Test Environment

- [ ] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
  - platform: ssd1306_i2c
    id: display_ssd1306
    model: "SSD1305 128x32"
    flip_x: true
    flip_y: true
    offset_x: 4
    offset_y: 0
    invert: false
    address: 0x3C
    rotation: 0°
    brightness: 1
    contrast: 1
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ssd1306_base/ssd1306_base.cpp [hunk 1]
```diff
@@ -69,25 +97,37 @@ void SSD1306::setup() {
...
+  this->command(SSD1306_COMMAND_NORMAL_DISPLAY | this->invert_);
+
+  // Disable scrolling mode (0x2E)
   this->command(SSD1306_COMMAND_DEACTIVATE_SCROLL);
 
-  set_brightness(this->brightness_);
+  // Contrast and brighrness
+  // SSD1306 does not have brightness setting
+  set_contrast(this->contrast_);
+  if (!this->is_ssd1305_())
```
Comment (jesserockz) [inline esphome/components/ssd1306_base/ssd1306_base.cpp]: Is this supposed to be only for ssd1305?
```suggestion
  if (this->is_ssd1305_())
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
If you are ignoring the brightness for ssd1306, then you might as well just `raise cv.Invalid` with a message to change. People miss/ignore warnings a lot and it's better to just get people to change the config with a breaking change in this case. 

Otherwise this PR becomes an unexpected change in functionality, instead of a breaking change on a simple config option to get the same functionality.

File: esphome/components/ssd1306_base/ssd1306_base.cpp (hunk 1 above)
Comment (zhangjingye03) [reply esphome/components/ssd1306_base/ssd1306_base.cpp]: Yes, a stupid mistake :-) Thanks for pointing out

**Comment by zhangjingye03** [comment]
> If you are ignoring the brightness for ssd1306, then you might as well just `raise cv.Invalid` with a message to change. People miss/ignore warnings a lot and it's better to just get people to change the config with a breaking change in this case.
> 
> Otherwise this PR becomes an unexpected change in functionality, instead of a breaking change on a simple config option to get the same functionality.

OK, I changed it to `raise cv.Invalid`.

File: esphome/components/ssd1306_base/__init__.py [hunk 2]
```diff
@@ -8,12 +9,21 @@
     CONF_MODEL,
     CONF_RESET_PIN,
     CONF_BRIGHTNESS,
+    CONF_CONTRAST,
+    CONF_INVERT,
 )
 
+_LOGGER = logging.getLogger(__name__)
+
 ssd1306_base_ns = cg.esphome_ns.namespace("ssd1306_base")
```
Comment (jesserockz) [inline esphome/components/ssd1306_base/__init__.py]: Not needed anymore
```suggestion

ssd1306_base_ns = cg.esphome_ns.namespace("ssd1306_base")
```

File: esphome/components/ssd1306_base/__init__.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import logging
```
Comment (jesserockz) [inline esphome/components/ssd1306_base/__init__.py]:
```suggestion
```

**Review by oxan — APPROVED** [review]

**Pull request merged**